compiler/src/dotty/tools/backend/jvm/BCodeBodyBuilder.scala

```diff
@@ -578,54 +621,39 @@ trait BCodeBodyBuilder extends BCodeSkelBuilder {
          * that cross cleanup boundaries. However, in theory such crossings are valid, so we should take care
          * of them.
          */
-        val resultKind = toTypeKind(fromSym.info)
-        genLoad(expr, resultKind)
-        lineNumber(r)
-        bc goTo programPoint(fromSym)
+        val (exprExpectedType, exprDest) = findJumpDest(fromSym)
+        genLoadTo(expr, exprExpectedType, exprDest)
       }
     } // end of genReturn()
 
-    def genWhileDo(tree: WhileDo, expectedType: BType): BType = tree match{
+    def genWhileDo(tree: WhileDo): LoadDestination = tree match{
       case WhileDo(cond, body) =>
 
       val isInfinite = cond == tpd.EmptyTree
 
+      val loop = new asm.Label
+      markProgramPoint(loop)
+
       if isInfinite then
-        body match
-          case Labeled(bind, expr) if tpeTK(body) == UNIT =>
-            // this is the shape of tailrec methods
-            val loop = programPoint(bind.symbol)
-            markProgramPoint(loop)
-            genLoad(expr, UNIT)
-            bc goTo loop
-          case _ =>
-            val loop = new asm.Label
-            markProgramPoint(loop)
-            genLoad(body, UNIT)
-            bc goTo loop
-        end match
-        expectedType
+        val dest = LoadDestination.Jump(loop)
+        genLoadTo(body, UNIT, dest)
+        dest
       else
         body match
           case Literal(value) if value.tag == UnitTag =>
             // this is the shape of do..while loops
-            val loop = new asm.Label
-            markProgramPoint(loop)
             val exitLoop = new asm.Label
             genCond(cond, loop, exitLoop, targetIfNoJump = exitLoop)
             markProgramPoint(exitLoop)
           case _ =>
-            val loop = new asm.Label
             val success = new asm.Label
             val failure = new asm.Label
-            markProgramPoint(loop)
             genCond(cond, success, failure, targetIfNoJump = success)
             markProgramPoint(success)
-            genLoad(body, UNIT)
-            bc goTo loop
+            genLoadTo(body, UNIT, LoadDestination.Jump(loop))
             markProgramPoint(failure)
         end match
-        UNIT
+        LoadDestination.FallThrough
```

Comment:
`genWhileDo` is the only method that returns a `LoadDestination`, and it constantly returns `FallThrough`, so maybe make the method `: Unit`?

Reply:
Right, I missed that. And the beginning of the `loop` destination then propagates down, which takes care of the tailrec handling.

Reply:
Yes, exactly. :)